File: expertAgent/aiagent/langgraph/jobGeneratorV2/workflows/workflow_gen/schemas/variable_patterns.py
"""TaskFlow V2 Variable Pattern Definitions.

Issue #351: Centralized variable pattern definitions for TaskFlow V2.

This module provides:
- TASKFLOW_VARIABLE_PATTERN: Compiled regex for variable references
- contains_variable_reference(): Check if value contains variables
- replace_variables_with_placeholder(): Replace variables for validation
- mask_secret_references(): Mask ${secrets.*} for logging
- validate_variable_syntax(): Validate variable syntax
"""

from __future__ import annotations

import re

# TaskFlow V2 variable reference pattern (MF-1: Unified pattern with hyphen support)
# Matches: ${inputs.query}, ${step_001.output}, ${step-001.output.data.name}, ${secrets.API_KEY}
# Pattern breakdown:
#   \$\{                           - Literal ${
#   [a-zA-Z_][a-zA-Z0-9_-]*        - Identifier (starts with letter/underscore, allows hyphen)
#   (?:\.[a-zA-Z_][a-zA-Z0-9_-]*)* - Optional dot-separated nested identifiers
#   \}                             - Literal }
TASKFLOW_VARIABLE_PATTERN = re.compile(
    r"\$\{[a-zA-Z_][a-zA-Z0-9_-]*(?:\.[a-zA-Z_][a-zA-Z0-9_-]*)*\}"
)

# Placeholder used during JSON validation (without quotes since variable refs are inside strings)
_VALIDATION_PLACEHOLDER = "__TASKFLOW_VAR_PLACEHOLDER__"


def contains_variable_reference(value: str) -> bool:
    """Check if value contains TaskFlow variable references.

    Args:
        value: String to check

    Returns:
        True if value contains ${...} variable references

    Example:
        >>> contains_variable_reference('{"result": "${step.output}"}')
        True
        >>> contains_variable_reference('{"result": "static"}')
        False
    """
    return bool(TASKFLOW_VARIABLE_PATTERN.search(value))


def replace_variables_with_placeholder(value: str) -> str:
    """Replace variable references with valid JSON placeholder.

    Used during validation to check JSON structure while allowing variables.

    Args:
        value: JSON string potentially containing variable references

    Returns:
        String with variables replaced by placeholder

    Example:
        >>> replace_variables_with_placeholder('{"result": "${step.output}"}')
        '{"result": "__TASKFLOW_VAR_PLACEHOLDER__"}'
    """
    return TASKFLOW_VARIABLE_PATTERN.sub(_VALIDATION_PLACEHOLDER, value)


def mask_secret_references(value: str) -> str:
    """Mask secret references for safe logging.

    Replaces ${secrets.KEY_NAME} with ${secrets.***} to prevent
    secret key names from appearing in logs.

    Args:
        value: String potentially containing secret references

    Returns:
        String with secret references masked

    Example:
        >>> mask_secret_references('Bearer ${secrets.API_TOKEN}')
        'Bearer ${secrets.***}'
    """
    return re.sub(r"\$\{secrets\.[^}]+\}", "${secrets.***}", value)


def validate_variable_syntax(value: str) -> list[str]:
    """Validate that all ${...} patterns have valid variable syntax.

    Returns list of invalid variable references found. Empty list means
    all variable references are valid.

    Args:
        value: String to validate

    Returns:
        List of invalid variable reference strings

    Example:
        >>> validate_variable_syntax('${valid.ref} and ${123invalid}')
        ['${123invalid}']
    """
    # Find all ${...} patterns (including potentially invalid ones)
    all_refs = re.findall(r"\$\{([^}]*)\}", value)
    invalid = []
    valid_pattern = re.compile(
        r"^[a-zA-Z_][a-zA-Z0-9_-]*(?:\.[a-zA-Z_][a-zA-Z0-9_-]*)*$"
    )
    for ref in all_refs:
        if not valid_pattern.match(ref):
            invalid.append(f"${{{ref}}}")
    return invalid


__all__ = [
    "TASKFLOW_VARIABLE_PATTERN",
    "contains_variable_reference",
    "replace_variables_with_placeholder",
    "mask_secret_references",
    "validate_variable_syntax",
]

File: graphAiServer/src/engine/constants/variable-patterns.ts
/**
 * TaskFlow V2 Variable Pattern Definitions
 *
 * Issue #352: Centralized variable pattern definitions for GraphAiServer.
 * Aligned with expertAgent/variable_patterns.py for consistency.
 *
 * This module provides:
 * - TASKFLOW_VARIABLE_PATTERN: Regex for variable references
 * - isValidTaskflowVariable(): Validate variable reference syntax
 * - extractVariableReference(): Extract variable from URL with trailing path
 *
 * Variable Reference Format:
 * - ${inputs.field} - Workflow input parameters
 * - ${step_id.output.field} - Previous step output
 * - ${env.VAR_NAME} - Environment variable
 * - ${secrets.KEY} - MyVault secret
 *
 * @module engine/constants/variable-patterns
 * @see Issue #352
 */

// ============================================================
// Pattern Definition
// ============================================================

/**
 * TaskFlow V2 variable reference pattern
 *
 * Pattern breakdown:
 *   \$\{                           - Literal ${
 *   [a-zA-Z_][a-zA-Z0-9_-]*        - Identifier (starts with letter/underscore, allows hyphen)
 *   (?:\.[a-zA-Z_][a-zA-Z0-9_-]*)* - Optional dot-separated nested identifiers
 *   \}                             - Literal }
 *
 * Matches:
 * - ${inputs.query}
 * - ${step_001.output}
 * - ${step-001.output.data.name}
 * - ${secrets.API_KEY}
 * - ${env.BASE_URL}
 *
 * Does NOT match:
 * - ${123invalid} - starts with number
 * - ${.invalid} - starts with dot
 * - ${} - empty
 * - ${invalid.} - ends with dot
 */
export const TASKFLOW_VARIABLE_PATTERN = /^\$\{[a-zA-Z_][a-zA-Z0-9_-]*(?:\.[a-zA-Z_][a-zA-Z0-9_-]*)*\}$/;

/**
 * Pattern to extract variable reference from the beginning of a string
 * Useful for URLs like "${inputs.url}/api/v1/endpoint"
 */
export const TASKFLOW_VARIABLE_EXTRACT_PATTERN = /^\$\{[^}]+\}/;

// ============================================================
// Validation Functions
// ============================================================

/**
 * Check if a string is a valid TaskFlow variable reference
 *
 * @param str - String to validate (must be exactly the variable, e.g., "${inputs.url}")
 * @returns true if the string is a valid variable reference
 *
 * @example
 * isValidTaskflowVariable("${inputs.url}") // true
 * isValidTaskflowVariable("${step_001.output.data}") // true
 * isValidTaskflowVariable("${123invalid}") // false
 * isValidTaskflowVariable("${inputs.url}/path") // false (has trailing path)
 */
export function isValidTaskflowVariable(str: string): boolean {
  return TASKFLOW_VARIABLE_PATTERN.test(str);
}

/**
 * Extract the variable reference from a string that may have trailing content
 * Returns null if no valid variable reference is found at the start
 *
 * @param str - String that may contain a variable reference at the start
 * @returns The variable reference or null
 *
 * @example
 * extractVariableReference("${inputs.url}/api/v1") // "${inputs.url}"
 * extractVariableReference("${env.BASE_URL}?query=1") // "${env.BASE_URL}"
 * extractVariableReference("https://example.com") // null
 */
export function extractVariableReference(str: string): string | null {
  const match = str.match(TASKFLOW_VARIABLE_EXTRACT_PATTERN);
  return match ? match[0] : null;
}

/**
 * Check if a URL starts with a valid TaskFlow variable reference
 * Supports trailing paths like "${inputs.url}/api/v1/endpoint"
 *
 * @param url - URL string to validate
 * @returns true if the URL starts with a valid variable reference
 *
 * @example
 * startsWithValidVariable("${inputs.url}") // true
 * startsWithValidVariable("${inputs.url}/api/v1") // true
 * startsWithValidVariable("${step_001.output.api_url}?query=1") // true
 * startsWithValidVariable("${123invalid}/path") // false
 * startsWithValidVariable("https://example.com") // false
 */
export function startsWithValidVariable(url: string): boolean {
  if (!url.startsWith('${')) {
    return false;
  }

  const varRef = extractVariableReference(url);
  if (!varRef) {
    return false;
  }

  return isValidTaskflowVariable(varRef);
}

// ============================================================
// Error Message Templates
// ============================================================

/**
 * Error message for invalid URL with helpful examples
 * Issue #352 SF-2: Improved error messages
 */
export const URL_VALIDATION_ERROR_MESSAGE =
  'URL must use HTTPS protocol or start with a TaskFlow variable reference. ' +
  'Valid variable formats: ${inputs.field}, ${step_id.output.field}, ${env.VAR}, ${secrets.KEY}. ' +
  'Example: ${inputs.base_url}/api/v1/endpoint';

/**
 * Short error message for Zod schema
 */
export const URL_VALIDATION_SHORT_MESSAGE =
  'URL must use HTTPS protocol or be a TaskFlow variable reference ' +
  '(${inputs.*}, ${step_id.output.*}, ${env.*}, ${secrets.*})';
